### PR TITLE
fix(turbopack): catchall route matching

### LINKED ISSUE #59502: `[...catchall]` routes do not work with turbopack
### Link to the code that reproduces this issue

https://github.com/Fredkiss3/turbopack-parallel-routes-bug

### To Reproduce

1.  Start the dev server with turbopack : `npm run dev -- --turbo`
2. Open the home page and navigates to`/other` using the links on the header, you will see that only the content of the parallel route in `@parallel/page.tsx` is shown when navigating
3. While on `/other` reload the page, the content will disappear, not even default.tsx is shown.

### Current vs. Expected behavior

Turbopack doesn't consider `[...catchAll]` routes and instead renders an empty content, while it works fine on webpack.

## Current behavior :
https://github.com/vercel/next.js/assets/38298743/d23fdde6-e687-445d-9a1c-6f94f98da9fe

## Expected behavior :
https://github.com/vercel/next.js/assets/38298743/d82d46fc-8fff-42d9-bd88-c0a37e64b251

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:45 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6020
Binaries:
  Node: 18.17.1
  npm: 9.6.7
  Yarn: 1.22.19
  pnpm: 8.7.1
Relevant Packages:
  next: 14.0.5-canary.6
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

_No response_

<sub>[NEXT-2038](https://linear.app/vercel/issue/NEXT-2038/[catchall]-routes-do-not-work-with-turbopack)</sub>

### PR DESCRIPTION
**Pull request by ForsakenHarmony**
This PR adds catch-all matching support to Turbopack. This allows you to have catch-all parallel slots. 

Closes NEXT-2038
Closes PACK-2518

Fixes #59502

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| buildDuration | 19.8s | 20s | ⚠️ +141ms |
| buildDurationCached | 7.7s | 7.9s | ⚠️ +137ms |
| nodeModulesSize | 196 MB | 196 MB | ✓ |
| nextStartRea..uration (ms) | 428ms | 420ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| 1068-HASH.js gzip | 30.2 kB | 30.2 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 4944-HASH.js gzip | 5.04 kB | 5.03 kB | N/A |
| 8423.HASH.js gzip | 181 B | 181 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 242 B | N/A |
| main-HASH.js gzip | 32.1 kB | 32.1 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 47.1 kB | 47.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 196 B | ✓ |
| _error-HASH.js gzip | 184 B | 183 B | N/A |
| amp-HASH.js gzip | 503 B | 504 B | N/A |
| css-HASH.js gzip | 323 B | 324 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.51 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 353 B | 351 B | N/A |
| hooks-HASH.js gzip | 370 B | 370 B | ✓ |
| image-HASH.js gzip | 4.21 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.68 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 313 B | 314 B | N/A |
| script-HASH.js gzip | 386 B | 385 B | N/A |
| withRouter-HASH.js gzip | 309 B | 311 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 931 B | 931 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 525 B | 524 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.4 kB | 94.5 kB | N/A |
| page.js gzip | 151 kB | 151 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 44.4 kB | 44.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.94 kB | 1.94 kB | ✓ |
| Overall change | 2.72 kB | 2.72 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 166 kB | 166 kB | ✓ |
| app-page-exp..prod.js gzip | 95.6 kB | 95.6 kB | ✓ |
| app-page-tur..prod.js gzip | 97.3 kB | 97.3 kB | ✓ |
| app-page-tur..prod.js gzip | 91.8 kB | 91.8 kB | ✓ |
| app-page.run...dev.js gzip | 136 kB | 136 kB | ✓ |
| app-page.run..prod.js gzip | 90.3 kB | 90.3 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.7 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 50 kB | 50 kB | ✓ |
| Overall change | 925 kB | 925 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/next-2038-turbopack-catchall-routes | Change |
| - | - | - | - |
| 0.pack gzip | 1.55 MB | 1.55 MB | N/A |
| index.pack gzip | 104 kB | 104 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

</details>
Commit: ff554c099d7c2ce3d09647fe8ee0f78d567201a2

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e0b433dbdeb7aa82dccc8d9f3527711f21ed9227

`TURBOPACK=1 pnpm test-dev test/e2e/app-dir/interception-routes-root-catchall/interception-routes-root-catchall.test.ts` (turbopack)
- interception-routes-root-catchall > should support having a root catch-all and a catch-all in a parallel route group

<details>
<summary>Expand output</summary>

  ● interception-routes-root-catchall › should support having a root catch-all and a catch-all in a parallel route group

    TIMED OUT: /Regular Item Page. Id: 1/

    Root Catch-All Page

    undefined

      636 |
      637 |   if (hardError) {
    > 638 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      639 |   }
      640 |   return false
      641 | }

      at check (lib/next-test-utils.ts:638:11)
      at Object.<anonymous> (e2e/app-dir/interception-routes-root-catchall/interception-routes-root-catchall.test.ts:23:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next-swc/crates/next-core/src/app_structure.rs [hunk 1]
```diff
@@ -433,6 +433,28 @@ impl LoaderTree {
...
+
+    /// Returns whether or not the only match in this tree is for a catch-all
+    /// route.
+    #[turbo_tasks::function]
+    pub async fn has_only_catchall(&self) -> Result<Vc<bool>> {
+        if self.segment == "__PAGE__"
+            && !matches!(
+                self.page.0.last(),
+                Some(PageSegment::CatchAll(..) | PageSegment::OptionalCatchAll(..))
+            )
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/next-core/src/app_structure.rs:445]: Maybe add `is_catchall` to `AppPage` as well

File: packages/next-swc/crates/next-core/src/app_structure.rs [hunk 2]
```diff
@@ -798,22 +820,29 @@ async fn directory_tree_to_loader_tree(
...
-                    app_dir,
-                    message: StyledString::Text(format!(
-                        "You cannot have two parallel pages that resolve to the same path. Route \
-                         {} has multiple matches in {}",
-                        for_app_path, app_page
-                    ))
-                    .cell(),
-                    severity: IssueSeverity::Error.cell(),
+            if let Some(current_tree) = tree.parallel_routes.get("children") {
+                if is_current_directory_catchall && *subtree.has_only_catchall().await? {
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/next-core/src/app_structure.rs:824]:
```suggestion
                if *subtree.has_only_catchall().await? {
```
Or is there a reason this is necessary?

File: packages/next-swc/crates/next-core/src/next_app/mod.rs [hunk 3]
```diff
@@ -208,6 +208,13 @@ impl AppPage {
         matches!(self.0.last(), Some(PageSegment::PageType(..)))
     }
 
+    pub fn is_catchall(&self) -> bool {
+        matches!(
+            self.0.last(),
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/next-core/src/next_app/mod.rs:213]: not necessarily correct as the completed page will have `PageSegment::PageType` as the last element (and catchall as potentially the second to last)

File: packages/next-swc/crates/next-core/src/app_structure.rs (hunk 2 above)
Comment (ztanner) [reply packages/next-swc/crates/next-core/src/app_structure.rs:824]: A handful of tests fail if I remove this, but I'll need to dig in a bit more to understand why.

**Review by timneutkens — APPROVED** [review]

**Pull request merged**